xc7/utils/prjxray_create_place_constraints.py

```diff
@@ -159,9 +159,12 @@ def __init__(self, conn, io_locs, blif_data):
 
                     clock['sink_nets'].append(sink_net)
 
-                    assert sink_net in self.input_pins or sink_net in self.clock_sources, (
-                        sink_net, self.input_pins, self.clock_sources.keys()
-                    )
+                    # TODO: Some sinks for clock networks could come from logic cells.
+                    # For instance, a negated clock does not have a clock as source, but
+                    # the output of a logic cell.
```

Comment:
This is (practically) never an acceptable solution for inverting a clock.  This should remain an error.  If a design is inverting a clock through logic, that is a **bug** in the synthesis.

Reply:
The problem is that we need an inverted clock, and the only way of obtaining it are:
- negate through logic (which is done by Vivado in the minitest as well)
- phasing by 180 degrees in the PLL (which does not actually work correctly in the minitest)
- using the MMCM that has CLKB outputs as well

Reply:
There is option 4:
 - Invert using dedicated inverters at the site

Where is this invert clock coming from and going to?

> * negate through logic (which is done by Vivado in the minitest as well)

Be more specific?  Is it inverted in logic the input netlist, or inverted in logic in the final placed/routed design.  If inverted in logic in the final P&R'd design, does that design actually met timing?

Reply:
> Where is this invert clock coming from and going to?

It comes from the fast clock (200MHz) clock buffer and it is routed to the CLKB pin of the ISERDESes

> Invert using dedicated inverters at the destination site

In this case this is not possible, at least now. I think that there is a `IS_CLKB_INVERTED` property, and if yes, we need to enhance the fuzzer to get support for it.

> Is it inverted in logic the input netlist, or inverted in logic in the final P&R'd design

I need to check this actually, Vivado automatically infers BUFGs for clock-related signals, in fact I could not find this clock inverted in logic in the top.edif file from the minitest.

Reply:
> It comes from the fast clock (200MHz) clock buffer and it is routed to the CLKB pin of the ISERDESes

The CLKB pin to the ISERDES **has** a dedicated invert.  Using a logic-based inverter here is a synthesis **bug**.

Reply:
![CLKB_inverter](https://user-images.githubusercontent.com/537074/73681588-85e85280-4673-11ea-8059-235583067a82.png)

Reply:
Then we need enhancement to the bufg fuzzer to be able to use that inverter. I'll also make sure it works in the minitest as well.

Reply:
> In this case this is not possible, at least now. I think that there is a `IS_CLKB_INVERTED` property, and if yes, we need to enhance the fuzzer to get support for it.

I'm not confident that this feature actually exists (or it may be conflated with ISERDES in use).

No matter the conclusion, using a logic-based clock inverter instead of the dedicate hardware clock inverter is never the right answer.  The timing violations are likely a **result** of using the logic-based clock inverter.

Reply:
> Then we need enhancement to the ILOGIC fuzzer to be able to use that inverter. I'll also make sure it works in the minitest as well.

Before going down this path, first make a minitest to explore what we want to do.  Given that mini-tests to date show no unknown bits, it is possible we've already encountered this bit.

Reply:
If we remove the change in this file, does the PR still go green?  I still believe this change is incorrect, and should be removed from this PR.

Reply:
Not yet, there still is the need to understand how to generate the inverted clock, or at least how to use the CLKB inverter in the ISERDESEs.

Reply:
@litghost this has been fixed. By using the same clock input for CLK and CLKB, the implementation works on HW.